### ISSUE TITLE
DisplayAPI

### ISSUE THREAD

**Issue by ksh-b** [termux/termux-api#595]
**Feature description**

API that would provide info about the display like screen size.

**Reference implementation**

Have you checked if the feature is accessible through the Android API?
> https://developer.android.com/reference/android/view/WindowMetrics

Do you know of other open-source apps that has a similar feature as the one you want? (Provide links)
> https://bitbucket.org/oF2pks/kaltura-device-info-android/src/master/

Output from the kaltura app:
```
    "#DISPLAY=": {
        "Built-inScreen_falseHDR": {
            "Name": "Built-in Screen",
            "State": "ON",
            "ro.sf.lcd_density": "300",
            "RefreshRate": 60.000003814697266,
            "ScreenSpecs": "720 x 1280",
            "ViewingMetrics": "DisplayMetrics{density=1.875, width=720, height=1280, scaledDensity=1.875, xdpi=294.967, ydpi=295.563}",
            "AllMetrics": "720 x 1280, largest app 1280 x 1280, smallest app 720 x 720, appVsyncOff 2000000, presDeadline 11666666, mode 1, defaultMode 1, modes [{id=1, width=720, height=1280, fps=60.000004, alternativeRefreshRates=[]}], hdrCapabilities HdrCapabilities{mSupportedHdrTypes=[], mMaxLuminance=500.0, mMaxAverageLuminance=500.0, mMinLuminance=0.0}, userDisabledHdrTypes [], minimalPostProcessingSupported false, rotation 0, state ON}",
            "SupportedHdrTypes": "",
            "SupportedModes": "{id=1, width=720, height=1280, fps=60.000004, alternativeRefreshRates=[]}",
            "Mode": "{id=1, width=720, height=1280, fps=60.000004, alternativeRefreshRates=[]}",
            "RefreshRates21": "60.000004",
            "Cutout": "null"
        }
    }
```

**Comment by Baba01hacker666** [comment]
I think they need to also add functionality like pip pictures in pictures and the ability to change system settings and access more WiFi information

**Comment by 0m364** [comment]
you don't need an API....
''' #!/bin/bash

# Get screen resolution
screen_size=$(dumpsys window | grep mUnrestrictedScreen | cut -d'=' -f2 | cut -d' ' -f1)
echo "Screen resolution: ${screen_size}"
'''
name that something . sh and $ bash your name.sh

**Comment by 0m364** [comment]
if you want a program here is one in kootlin

'''
 val displayMetrics = DisplayMetrics()
windowManager.defaultDisplay.getMetrics(displayMetrics)
val height = displayMetrics.heightPixels
val width = displayMetrics.widthPixels
'''

**Comment by 0m364** [comment]
get_screen_res.py
'''
import subprocess

def get_screen_resolution():
    result = subprocess.run(['dumpsys', 'window'], stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, check=True)
    lines = result.stdout.splitlines()

    for line in lines:
        if 'mUnrestrictedScreen' in line:
            screen_size = line.split('=')[1].split()[0]
            return screen_size

print(f"Screen resolution: {get_screen_resolution()}")
'''

**Comment by Rudxain** [comment]
@0m364 
```
~ $ dumpsys window
Can't find service: window
```
Android 12.0

That command requires `adb shell`, to work on all devices

**Comment by 0m364** [comment]
> @0m364 
> ```
> ~ $ dumpsys window
> Can't find service: window
> ```
> Android 12.0
> 
> That command requires `adb shell`, to work on all devices

you are correct I was working from my rooted device none of my commands work non root I am seeing now.. I will attempt to find a work around but since dumpsys is at a level termux doesn't access without root I'm not seeing a straightforward or easy way to do this... possibly by implementing something with x11 where it tries to determine screen size before providing a gui instead just display the values... I'll play around with it al little bit later... there should be a way to display screen sizes and frame rates

**Comment by Rudxain** [comment]
> dumpsys is at a level termux doesn't access without root 

`adb` can be installed with
```sh
pkg i android-tools
```
but the pairing+connecting process is a big pain. The pairing-port and the connection-port are different, and multi-window must be used for pairing.

> possibly by implementing something with x11 where it tries to determine screen size before providing a gui instead just display the values...

That's an interesting approach! I have no idea how to use the x11 API, but I guess it'll work

**Comment by Ammar-Husain** [comment]
none of the approaches you guys mention does work in unrooted devices, even dumpsys isn't available without root.
can't this feature be added to the api?

**Comment by Rudxain** [comment]
> none of the approaches you guys mention does work

Which Android version? Perhaps Google hardened the system again.

> can't this feature be added to the api?

[It's technically possible](https://llamalab.com/automate/doc/block/display_metrics_get.html)

**Comment by Biswa96** [comment]
I have added a pull request to fix this issue. Would you like to test the packages from GitHub Actions artifacts of that pull request?
